src/mlpack/methods/hmm/hmm_impl.hpp

```diff
@@ -620,6 +632,20 @@ void HMM<Distribution>::Backward(const arma::mat& dataSeq,
   // The last element probability is 1.
   backwardLogProb.col(dataSeq.n_cols - 1).fill(0);
 
+  // Define a variable to store the value of log-probability for dataSeq.
+  arma::mat logProbs(transition.n_rows, dataSeq.n_cols);
+  logProbs.fill(-std::numeric_limits<double>::infinity());
+
+  // Save the values of log-probability to logProbs.
+  for (size_t i = 0; i < transition.n_rows; i++)
+  {
+    // Define arma::vec to store log-probability for any dataSeq column.
+    arma::vec logProb;
+    emission[i].LogProbability(dataSeq, logProb);
+    for (size_t j = 0; j < logProb.n_elem; j++)
+      logProbs(i, j) = logProb(j);
```

Comment:
Actually we can be a little more tricky here.  Armadillo's `unsafe_col()` function allows us to create an `arma::vec` that has memory that points to a column of an existing matrix.  So you can do

```
emission[i].LogProbability(dataSeq, logProbs.unsafe_col(i))
```

However, I see that `logProbs` is of size `transition.n_rows` by `dataSeq.n_cols`.  But really you would want this to be transposed (`dataSeq.n_cols` by `transition.n_rows`) because Armadillo stores matrices in a column-major format.  That means it is a lot faster to access elements that are next to each other in a single column than in a single row.  Only after making that change will my suggestion above work.

In addition, when you do this, you won't need to use `fill()` or `zeros()`, since `LogProbability()` will set the elements to their desired value.

Reply:
@rcurtin , You were correct. There is a significant increase in the efficiency of the algorithm. The timing got reduced from 0.0027 to around 0.0016. Do you wish to add any other thing apart from the work already done?